### DISCUSSION TITLE
Functionality to check if GPS/Location is enabled on mobile

### DISCUSSION THREAD

**Discussion by ouijahija** [dotnet/maui#22245] [Ideas]
### Description

To clarify: I'm know how to request permission for coarse/fine location. I'm looking for a way to detect, if the location tracking on the device is enabled or not. ComptonAlvaro requested this as well [here](https://github.com/dotnet/maui/issues/14682) and I'm really not sure why this issue was marked as resolved. The offered solution solution (checking it via the permission API if something is possible/available) does not work, as checking the permission status via 
`await Permissions.CheckStatusAsync<Permissions.LocationWhenInUse>();` 
returns 
`PermissionStatus.Granted`
even if the devices location tracking is disabled (at least on android).

Checking if the Location Service is enabled by
`await Geolocation.GetLocationAsync()`
doesn't work either, as it throws an exception: `**Microsoft.Maui.ApplicationModel.FeatureNotEnabledException:** 'Location services are not enabled on device.'`

### Public API Changes

```
//Returns false if the Location is disabled (Android), Precise Location is Disabled (iOS) or the device has no location service
if (Geolocation.IsLocationEnabled)
{
      //Start tracking location etc.
}
else
{
     //Ask user to enable location on their device
} 
```

### Intended Use-Case

My App doesn't require location to be on to be used but I want remind users to enable it, as it increases the the functionality of my App. Also I only want to check for the user's location if the Location Service is enabled.

**Comment by cucumber-sp** [comment]
Would be nice to also have something like `RequestToEnableLocation`. Android definitely has popup for this, not sure about iOS and other platforms tho

**Comment by salvadorjesus** [comment]
The way I do is using:

` var location = await Geolocation.GetLocationAsync(geolocationRequest);`

This can throw a `PermissionException` as well as a `FeatureNotSupportedException`. The latter case throws when the device location feature is off, permissions notwithstanding. If that is the case, I ask the user if she wants to enable it and open app settings so she can turn it on.

However, this is not a good solution. This call can take a number of seconds and sometimes I can't block the app for that.

**Reply by ouijahija** [reply]
Yeah, I do the same and wrap it in a try/catch statement and run it on a separate thread to not block the main thread. If it catches an exception, I know the location feature is disabled. It feels really hacky and unecessarily convoluted.

**Comment by DigitalRadtke** [comment]
I'm gonna bump this "discussion". I'm using quotation marks here, because let's face it: This should be marked as an Issue, as it is clearly not working as intended. There's currently (at least on those android devices I've tested) no way to check if the Gps Feature is enabled/disabled other than to make a GeolocationRequest, wrap it in a trycatch statement and to assume that the feature is disabled if the request throws an error.
The offered solution of checking the LocationWhenInUse permission and seeing if the permission status is PermissionStatus.Disabled IS NOT WORKING. It returns PermissionStatus.Granted or PermissionStatus.Denied no matter if the Gps Feature is enabled or not.

**This is a basic feature, that should've been implemented way before deprecating Xamarin.Forms and forcing us to switch to Maui (such as many other vital and basic features, that are missing). At least I'd appreciate some statement from one of the developers, when we can expect a FIX to this ISSUE.**

**Comment by ghost** [comment]
Would love to see a fix for this too.

In the meantime, you could possibly catch the exception and do something within that catch-block. For example, show a popup to ask the user to manually turn on their location, or notify the user that location services are not supported on their device:

	try
	{
		await Geolocation.GetLocationAsync();
                // check if not null, do something with location data...
	}
	catch (PermissionException e)
	{
		// No permission, show popup to user, etc...
	}
	catch (FeatureNotEnabledException e)
	{
		// Location turned OFF, show popup to user, etc...
	}
	catch (FeatureNotSupportedException e)
	{
		// Location services not supported, show popup to user, etc...
	}
	catch (Exception e)
	{
		// Something else went wrong
	}

Note that in debug-mode, VisualStudio may still throw an exception. When running the app in release mode the exception is not thrown, and your catch-block code will be executed instead.
